### ISSUE TITLE
Nuget 4.7 works, but later ones do not (Service Fabric projects)

### ISSUE THREAD

**Issue by karann-msft** [NuGet/Home#7784]
@galmok commented on [Tue Feb 05 2019](https://github.com/NuGet/docs.microsoft.com-nuget/issues/1300)

We are using nuget to restore the packages of our projects, but when we upgraded nuget from version 4.7.1 to 4.8.1 or to 4.9.1, this restore must be run twice to do the work that 4.7.1 does. The newer nugets issue this warning (originating from the Service Fabric project, *.sfproj file):

D:\temp\telemetryperiodservice\TelemetryPeriodServiceSF\TelemetryPeriodServiceSF.sfproj(47,5): error : Unable to find the '..\packages\Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.6.7\build\Microsoft.VisualStudio.Azure.Fabric.Application.props' file. Please restore the 'Microsoft.VisualStudio.Azure.Fabric.MSBuild' Nuget package.

The first package restored is actually the package it complains about.

Using nuget 4.7.1, it seems the mentioned package is already restored which makes all the nuget restore put packages into the projects and not just into the packages folder.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 300985dd-9626-6e0a-d055-73dbbc3633a0
* Version Independent ID: 814d9eaa-25dc-58f2-e0be-a390f3d360b7
* Content: [NuGet 4.9 RTM Release Notes](https://docs.microsoft.com/en-us/nuget/release-notes/nuget-4.9-rtm)
* Content Source: [docs/release-notes/NuGet-4.9-RTM.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master/docs/release-notes/NuGet-4.9-RTM.md)
* Product: **nuget**
* GitHub Login: @karann-msft
* Microsoft Alias: **karann**

**Comment by silvioboehme** [comment]
Hi!

We have the same problem with nuget version 4.9.2 and 4.9.3.
This breaks out Azure DevOps build, and we have to use the old 4.7.1 version.

**Comment by zivkan** [comment]
Unfortunately this is due to a design mismatch between the default Service Fabric project template and how NuGet works.

### tl;dr

Edit your `.sfproj`, remove the `InitialTargets` attribute from the root `Project` node, and if you want add `BeforeTargets="Build"` to the `ValidateMSBuildFiles` target.

### Detailed explanation

Since NuGet added support for `ProjectReference`, we need to do an MSBuild evaluation of every project in the solution because `ProjectReferences` could be in included in an `Import` or implicitly though `Directory.Build.props` and in any case NuGet doesn't want to duplicate MSBuild parsing logic. For reasons I don't know, NuGet 4.7 and earlier appears to whitelist the file extensions that it will evaluate, but this prevents custom project types from using `PackageReference` unnecessarily, so NuGet 4.8 and onwards will evaluate every project type, regardless of extension, so it's not limited to .NET projects.

The default `sfproj` template has a target called `ValidateMSBuildFiles` which forces MSBuild to report an error when the files it expects from it's build tools nuget packages do not exist, which of course do not exist because it hasn't been restored yet. Unfortunately the template also adds this target as an `InitialTarget`, which instructs MSBuild to run this target, in addition to whatever target MSBuild was asked to run. So, NuGet asks MSBuild to run some custom targets on the project to figure out if there are any `PackageReference`, the project instructs MSBuild to always run its own target which will always force an error, so MSBuild reports an error to NuGet, so NuGet restore fails because it can't restore projects using PackageReference.

Without blacklisting `.sfproj` from MSBuild evaulations, and therefore preventing them from ever using `PackageReference`, there's nothing NuGet can really do about this. The Service Fabric tooling team are working towards making `.sfproj` use MSBuild SDK style files, but since only .NET projects currently use SDK on any significant scale, there are cross team issues that need to be worked on, so I don't imagine that they have made any claims when this will be available. I emailed the person I talked to when I first discovered this issue to ask if he could apply the tl;dr workaround above to the current template, until the SDK migration is ready. I can't make any promises about any outcome though.
